### ISSUE TITLE
"network is disconnected" forever after switching from ethernet to wifi

### ISSUE THREAD

**Issue by larsonmpdx** [aws/aws-iot-device-sdk-cpp#81]
using the general sync() workflow in #80 

with a good connection to aws iot, upon switching from ethernet to wifi (by unplugging the ethernet cable and waiting for networkmanager to figure things out), the aws iot connection fails and then after sync connect() is called with wifi available it connects.  but the first published message fails with:

`:326 [Publish] [3069726720] [PerformAction:L210] : Publish Write to Network Failed. Network is disconnected : SDK Code -500.`

the client application then loops and tries to connect again, is successful in connecting, and then fails, forever

it looks like some part the network code isn't getting reset after a failure.  how can I fix this?

**Comment by larsonmpdx** [comment]
I should add, auto reconnect with a sync disconnect then sync connect running in parallel handled this situation correctly but it was subject to deadlocks as in #80 

the new loop is just a sync connect() call and no auto reconnect or disconnect(), and this is how it's failing

**Comment by vareddy** [comment]
Hi @larsonmpdx,
I will try to reproduce this issue and get back to you. The test flow, from what I can gather from above is:
- connect to iot using sync connect
- start publish
- change network connection from ethernet to wifi
- (should reconnect and try to publish)
- fails forever
Is this correct?
Thanks!
Varun

**Comment by larsonmpdx** [comment]
yes, when your publish fails loop back to your sync connnect(), that should reproduce it

thanks!

**Comment by larsonmpdx** [comment]
Varun,

Thanks for trying to reproduce! I'm on Linux. I'll try to set up a similar reproducer in an Ubuntu VM. I think I'll get to this on Thursday

**Comment by larsonmpdx** [comment]
please see this for a version that demonstrates the problem on an ubuntu VM

https://github.com/larsonmpdx/aws-iot-device-sdk-cpp/blob/sync_connect_loop/samples/PubSub/PubSub.cpp#L244

I don't need to switch wifi/ethernet to demonstrate it, only to disable networking in the GUI, wait a moment, and then enable it again
![image](https://user-images.githubusercontent.com/12436438/38382805-c08623f4-38bf-11e8-8208-2288dc29f00b.png)

**Comment by larsonmpdx** [comment]
this also returns -703, not -500, but the effect is the same

**Comment by larsonmpdx** [comment]
do you need help reproducing this?  I think this is the problem affecting my fielded units as they encounter normal network problems and occasionally get permanently stuck

**Comment by vareddy** [comment]
Hi @larsonmpdx ,
Apologies for the delay, I haven't had time to look deeper into this issue. I will try your version of the reproducer and check if that helps.
When you say occasionally, does this happen after repeated connect-reconnect cycles or every time when it attempts a reconnect?  
One drastic change you could try is using a new network instance (or even a new MQTT instance along with a new network instance) every time it doesn't recover.  I haven't tried this but it could possibly be worth the overhead. 
Thanks!
Varun

**Comment by larsonmpdx** [comment]
I can reproduce this by forcing my device to switch from wifi to a cell internet connection.  It also happens randomly if the device is left alone on a cell connection, I assume this is because the cell network drops and restarts in an odd state sometimes.  This usually takes 1-3 days.

I will try recreating things within the loop, the device isn't resource constrained so overhead doesn't matter.  I tried variations of this while debugging and I suspected that I had some trouble with the destructors deadlocking but I didn't narrow things down.  It's worth another try, thank you.
